optimize redis for leakybucket and tokenbucket

leeym opened the pull request:
Currently in Redis we use different keys for different values, like Last and Available, and we use pipeline to make sure to update multiple keys in an atomic way. Given that we always access these two keys together, it will be easier to just use one document to store both values.

mennanov commented:
This is a great improvement! 
However i'm not sure how to rollout this change properly: the new version of this code will start reading and writing from a different Redis object which means that the already existing state will become forgotten. This may result in under-limiting the requests.

How would you approach it?

leeym commented:
> How would you approach it?

I did the following to make sure it works as expected
- renamed existing `State` which reads from multiple keys to `oldState`
- added a new `State` to read from one key with multiple values, and fall back to `oldState` if needed
- renamed existing `SetState` to `SetStateForTesting` which writes multiple keys
- added a new `SetState` to write only one key with multiple values
- added unit tests to write with `SetStateForTesting` and read it by `State` to make sure it is backward compatible

mennanov commented on leakybucket.go:
```diff
@@ -379,9 +504,9 @@ func checkResponseFromRedis(response interface{}, expected interface{}, expected
+// Deprecated: SetStateForTesting updates the state in Redis.
 // The provided fencing token is checked on the Redis side before saving the keys.
-func (t *LeakyBucketRedis) SetState(ctx context.Context, state LeakyBucketState) error {
+func (t *LeakyBucketRedis) SetStateForTesting(ctx context.Context, state LeakyBucketState) error {
```
Is it possible to make this method private so that it's only available in tests?

leeym replied:
I moved that into the test file, and set the key using redisClient directly.

mennanov reviewed (approved):
LGTM!

What do you think is the right way to release this change? 
I'm thinking of the following: release this PR as a patch version update and release the next PR which removes the deprecated methods as a minor version update.

leeym commented:
> I'm thinking of the following: release this PR as a patch version update and release the next PR which removes the deprecated methods as a minor version update.

We might want to keep them both for a while. For example, as of now the latest version is v1.13.0, and you release this feature in v1.13.1, but remove the deprecated code in v1.14.0. It means that it only works if the user follows closely along the way: v1.13.0 -> v1.13.1 -> v1.14.0. Any one jumps directly from v1.13.0 or older to v1.14.0 or newer will feel that the old keys are gone suddenly.

We can keep both features for long enough, for example a few months or even a full year, before we remove the deprecated one. We can even leave it as is until there is a breaking change and remove deprecated code altogether.

mennanov commented:
> We might want to keep them both for a while. For example, as of now the latest version is v1.13.0, and you release this feature in v1.13.1, but remove the deprecated code in v1.14.0. It means that it only works if the user follows closely along the way: v1.13.0 -> v1.13.1 -> v1.14.0. Any one jumps directly from v1.13.0 or older to v1.14.0 or newer will feel that the old keys are gone suddenly.
> 
> We can keep both features for long enough, for example a few months or even a full year, before we remove the deprecated one. We can even leave it as is until there is a breaking change and remove deprecated code altogether.

Makes sense, let's merge the future PR with the clean-up  in a few months.

Feel free to merge this PR
